File: R/gradient-descent/multiple-regression.R
# # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#               File Owner - Kartheek Palepu            #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # #

# # Loss function = half-MSE # #
# # minimize the loss

loss = function(y, yhat) {
	se = sum((yhat - y) ^ 2)
	mse = se / (2.0 * length(y))
	return(mse)
}

derivative1 = function(y, yhat, bias = 1) {
	d = sum(yhat - y)
	d = d * (bias / length(y)) # let bias term = 1
	return(d)
}

derivativeN = function(y, yhat, x_n, bias = 1) {
	d = sum((yhat - y) * x_n)
	d = d * (bias / length(y)) # let bias term = 1
	return(d)
}

yEQ = function(x, intercept, slopes) {
	mx = as.matrix(x) %*% diag(slopes)
	mxc = rowSums(mx) + intercept
	return(mxc)
}


# # # Define the hyper parameters
# # learning rate
alpha = 0.001
# # threshold error rate
threshold = 0.0001

# # generate x
x1 = 1:10
x2 = seq(1, 20, 2)
x3 = seq(0, 19, 2)
x = data.frame(x1, x2, x3)

# # formulate original 'y' i.e. y = 1 + x1 + 2*x2 + x3
slopes = c(1, 2, 1)
intercept = 1
y = yEQ(x, intercept = intercept, slopes = slopes)

# # initialize values
slopes = rep(0, ncol(x))
intercept = 0

# # start iterations
for(i in 1:10000){
	yhat = yEQ(x, intercept = intercept, slopes = slopes)
	if(loss(y, yhat) <= threshold) {
		print("==============================================================================")
		print("==============================================================================")
		print(paste0("== Converged at ", i, " =="))
		print(paste0("- Slopes = ", paste(slopes, collapse = ", ")))
		print(paste0("- Intercept = ", intercept))
		print(paste0("- Y values = ", paste(y, collapse = ", ")))
		print(paste0("- Yhat values = ", paste(yhat, collapse = ", ")))
		print(paste0("- Loss = ", loss(y, yhat)))
		break;
	}
	if((i - 1) %% 100 == 0){
		print(paste0("At iteration #: ", i, "- At Slopes [", paste(slopes, collapse = ", "), "] and intercept = ", intercept, " loss is ", loss(y, yhat)))
	}
	intercept = intercept - alpha * derivative1(y, yhat)
	for(j in 1:length(slopes)){
		slopes[j] = slopes[j] - alpha * derivativeN(y, yhat, x[j])
	}
}

File: Python/grandient-descent/multiple-regression.py
#!/usr/bin/env python
# # # # # # # # # # # # # # # # # # # # # # # # # # # # #
#               File Owner - Kartheek Palepu            #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # #

# # Loss function = half-MSE # #
# # minimize the loss
def loss(y, yhat):
	se = round(sum([(yhat[i] - y[i])**2 for i in range(len(y))]), 6)
	mse = se/(2.0 * len(y))
	return mse

def derivative1(y, yhat):
	d = sum([yhat[i] - y[i] for i in range(len(y))])
	d = d / float(len(y))
	return(d)

def derivativeN(y, yhat, x_n):
	d = sum([((yhat[i] - y[i]) * x_n[i]) for i in range(len(y))])
	d = d / float(len(y))
	return(d)

def yEQ(x, intercept, slopes):
	mx = [[j * slopes[i] for j in x[i]] for i in range(len(x))]
	mxc = [intercept + sum(i) for i in zip(*mx)]
	return mxc


# # # Define the hyper parameters
# # learning rate
alpha = 0.001
# # threshold error rate
threshold = 0.0001

# # generate x
x = []
x1 = range(1, 11)
x2 = range(1, 20, 2)
x3 = range(0, 19, 2)
x.append(x1)
x.append(x2)
x.append(x3)

# # formulate original 'y' i.e. y = 1 + x1 + 2*x2 + x3
slopes = [1, 2, 1]
intercept = 1
y = yEQ(x, intercept = intercept, slopes = slopes)

# # initialize values
slopes = [0] * len(x)
intercept = 0

# # start iterations
for i in range(10000):
	print "ITERATION # " + str(i)
	yhat = yEQ(x, intercept = intercept, slopes = slopes)
	if loss(y, yhat) <= threshold:
		print "=============================================================================="
		print "=============================================================================="
		print "== Converged at " + str(i) + " =="
		print "- Slopes = " + ",".join([str(k) for k in slopes])
		print "- Intercept = " + str(intercept)
		print "- Y values = " + ", ".join([str(k) for k in y])
		print "- Yhat values = " + ", ".join([str(k) for k in yhat])
		print "- Loss = " + str(loss(y, yhat))
		break;
	print "At Slopes [" + ", ".join([str(k) for k in slopes]) + "] and intercept = " + str(intercept) + " loss is " + str(loss(y, yhat))
	intercept = round(intercept - alpha * derivative1(y, yhat), 6)
	for j in range(len(slopes)):
		slopes[j] = round(slopes[j] - alpha * derivativeN(y, yhat, x[j]), 6)
